docs11/man/man7/Controller.7
.TH "RJFRC2011::Controller" 7 "Mon May 30 2011" "RJHS FRC Software Documentation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
RJFRC2011::Controller \- 
.PP
Class abstracting the controller(s) used by our driver(s)  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Controller.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBController\fP ()"
.br
.RI "\fIConstructor. \fP"
.ti -1c
.RI "\fB~Controller\fP ()"
.br
.RI "\fIDestructor. Kill stuff. \fP"
.ti -1c
.RI "float \fBgetManipulatorElevation\fP ()"
.br
.RI "\fIGet user-requested manipulator elevation. \fP"
.ti -1c
.RI "float \fBgetDriveSpeed\fP ()"
.br
.RI "\fIGet user-requested drive speed. \fP"
.ti -1c
.RI "float \fBgetDriveTurn\fP ()"
.br
.RI "\fIGet user-requested drive turn speed. \fP"
.ti -1c
.RI "int \fBgetMinibotSwitches\fP ()"
.br
.RI "\fIGet user-requested minibot shelf action (in/out). \fP"
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static int \fBgetManipulatorAction\fP ()"
.br
.RI "\fIGet user-requested manipulator action (input, rotate, or eject). \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "float \fBabs\fP (float initial)"
.br
.RI "\fIAbsolute value of a float, since I'm not sure if we can import the <cmath> library onto the cRIO. EDIT: turns out we can, but I'll leave this here to conserve memory. \fP"
.ti -1c
.RI "float \fBexpo\fP (float x, float a)"
.br
.RI "\fIAn exponential function used to make joysticks less sensitive near the center and more sensitive towards the edges. \fP"
.ti -1c
.RI "float \fBnormalize\fP (float joyVal, float min, float max)"
.br
.RI "\fIA function that 'normalizes' inputs from the joysticks (because they don't give perfect -1.0 to 1.0 values). \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "Joystick * \fB_controller\fP"
.br
.RI "\fIObject abstracting our InterLink Elite controller, which WIPLib seems to think is a joystick. \fP"
.ti -1c
.RI "Joystick * \fB_joystick\fP"
.br
.RI "\fIObject abstracting our legitimate Logitech Attack3 joystick. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Class abstracting the controller(s) used by our driver(s) 

Get input from a pair of joysticks (one of which is physically a flight simulator controller) and return those inputs. A class like this is useful because we can modify a few lines of code to change the inputs for functions throughout the entire code. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "RJFRC2011::Controller::Controller ()"
.PP
Constructor. Basically, initialize the controllers to USB ports 1 and 2. 
.SH "Member Function Documentation"
.PP 
.SS "float RJFRC2011::Controller::abs (floatinitial)\fC [private]\fP"
.PP
Absolute value of a float, since I'm not sure if we can import the <cmath> library onto the cRIO. EDIT: turns out we can, but I'll leave this here to conserve memory. \fBParameters:\fP
.RS 4
\fIinitial\fP the initial value 
.RE
.PP
\fBReturns:\fP
.RS 4
the absolute value of the passed value; if it's negative, make it positive 
.RE
.PP
\fBAuthor:\fP
.RS 4
Matthew Haney 
.RE
.PP

.SS "float RJFRC2011::Controller::expo (floatx, floata)\fC [private]\fP"
.PP
An exponential function used to make joysticks less sensitive near the center and more sensitive towards the edges. Basically, plug the value requested by the user and a predefined constant into an exponential equation and return the result. 
.PP
\fBParameters:\fP
.RS 4
\fIx\fP the value to be exponentiated 
.br
\fIa\fP a predefined exponential factor 
.RE
.PP
\fBReturns:\fP
.RS 4
the 'expo-ed' value 
.RE
.PP
\fBAuthor:\fP
.RS 4
Adam Bryant 
.RE
.PP

.SS "float RJFRC2011::Controller::getDriveSpeed ()"
.PP
Get user-requested drive speed. Reads input from the y-axis on the right stick, normalizes it to a value between -1.0 and 1.0, and then exponentiates it (so the stick's less sensitive in the center and more sensitive at the edges). 
.PP
\fBReturns:\fP
.RS 4
The received, normalized, and expo-ed value (inverted [if the inversion switch {button #2} is thrown]). 
.RE
.PP

.SS "float RJFRC2011::Controller::getDriveTurn ()"
.PP
Get user-requested drive turn speed. Reads input from the x-axis on the right stick, normalizes it to a value between -1.0 and 1.0, and then exponentiates it (so the stick's less sensitive in the center and more sensitive at the edges). 
.PP
\fBReturns:\fP
.RS 4
The received, normalized, and expo-ed value(inverted [if the inversion switch {button #2} is thrown]). 
.RE
.PP

.SS "Controller::getManipulatorAction ()\fC [static]\fP"
.PP
Get user-requested manipulator action (input, rotate, or eject). Read input from three buttons on the joystick: the trigger (1) and two thumb buttons (2 and 3). If any one of them is pressed, return that button's ID. If multiple are pressed, or none are pressed, return 0. 
.PP
\fBReturns:\fP
.RS 4
The ID of the received button. 
.RE
.PP

.SS "float RJFRC2011::Controller::getManipulatorElevation ()"
.PP
Get user-requested manipulator elevation. Reads input from the y-axis on the joystick. 
.PP
\fBReturns:\fP
.RS 4
The received value (inverted). 
.RE
.PP

.SS "int RJFRC2011::Controller::getMinibotSwitches ()"
.PP
Get user-requested minibot shelf action (in/out). Read input from two thumb buttons on the joystick (buttons 4 and 5) and return ther states in a single variable. 
.PP
\fBReturns:\fP
.RS 4
An integer with the last two bits being the right and left button inputs, respectively. 
.RE
.PP

.SS "float RJFRC2011::Controller::normalize (floatjoyVal, floatmin, floatmax)\fC [private]\fP"
.PP
A function that 'normalizes' inputs from the joysticks (because they don't give perfect -1.0 to 1.0 values). If the requested value is negative, return its percentage of the minimum possible value; if it's possible, do the same with the max. If it's zero, of course, return zero. 
.PP
\fBParameters:\fP
.RS 4
\fIjoyVal\fP the input from the joystick 
.br
\fImin\fP the minimun joystick value 
.br
\fImax\fP the maximum joystick value 
.RE
.PP
\fBReturns:\fP
.RS 4
the normalized value 
.RE
.PP
\fBAuthor:\fP
.RS 4
Adam Bryant 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for RJHS FRC Software Documentation from the source code.


workspace11/BuiltinDefaultCode/Controller.cpp
/*
 RJFRC - The code used by Regis Jesuit High School's FIRST Robotics Competition team #3729 over the years, starting from 2011 (rookie year).
 Copyright (C) 2011 Regis Jesuit High School
 
 This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*!
 \file Controller.cpp
 \brief File containing implementations of functions found in the \a Controller class (found in Controller.h)
 \authors Matthew Haney, Drew Lazzeri
 
 */

#include "Controller.h"
#include "WPILib.h"
#include "macros.h"

using namespace RJFRC2011;

/*!
 \brief Constructor.
 \details Basically, initialize the controllers to USB ports 1 and 2.
 */
RJFRC2011::Controller::Controller()
{
	this->_controller = new Joystick(1);
	this->_joystick = new Joystick(2);
}

/*!
 \brief Destructor.  Kill stuff.
 */
RJFRC2011::Controller::~Controller()
{
	delete _controller;
	delete _joystick;
}

/*!
 \brief Get user-requested manipulator elevation.
 \details Reads input from the y-axis on the joystick.
 \return The received value (inverted).
 */
float RJFRC2011::Controller::getManipulatorElevation()
{
	//The below was used when we still had this input on the left stick of the InterLink Elite
	//return normalize(this->_controller->GetZ(), ZMIN, ZMAX) * -1.0;
	return this->_joystick->GetY() * -1.0;
}

/*!
 \brief Get user-requested drive speed.
 \details Reads input from the y-axis on the right stick, normalizes it to a value between -1.0 and 1.0, and then exponentiates it (so the stick's less sensitive in the center and more sensitive at the edges).
 \return The received, normalized, and expo-ed value (inverted [if the inversion switch {button #2} is thrown]).
 */
float RJFRC2011::Controller::getDriveSpeed()
{
	// Y-axis right; speed
	if (this->_controller->GetRawButton(2) == 1)
		return expo(normalize(this->_controller->GetY(), YMIN, YMAX) * -1.0, YEXPO) * -1.0;
	else
		return expo(normalize(this->_controller->GetY(), YMIN, YMAX) * -1.0, YEXPO);
}

/*!
 \brief Get user-requested drive turn speed.
 \details Reads input from the x-axis on the right stick, normalizes it to a value between -1.0 and 1.0, and then exponentiates it (so the stick's less sensitive in the center and more sensitive at the edges).
 \return The received, normalized, and expo-ed value(inverted [if the inversion switch {button #2} is thrown]).
 */
float RJFRC2011::Controller::getDriveTurn()
{
	// X-axis right; turn
	if (this->_controller->GetRawButton(2) == 1)
		return expo(normalize(this->_controller->GetX(), XMIN, XMAX), XEXPO) * -1.0;
	else
		return expo(normalize(this->_controller->GetX(), XMIN, XMAX), XEXPO);
}

/*!
 \brief Get user-requested manipulator action (input, rotate, or eject).
 \details Read input from three buttons on the joystick: the trigger (1) and two thumb buttons (2 and 3).  If any one of them is pressed, return that button's ID.  If multiple are pressed, or none are pressed, return 0.
 \return The ID of the received button.
 */
int RJFRC2011::Controller::getManipulatorAction()
{
	// X-axis left; manipulator receive/eject controls
	//return normalize(this->_controller->GetRawAxis(5), ROTMIN, ROTMAX);
	// 0 = none
	// 1 (trigger) = eject
	// 2 = rotate down
	// 3 = suck in
	if ((this->_joystick->GetRawButton(3) == 1 && this->_joystick->GetRawButton(1) == 1) || (this->_joystick->GetRawButton(3) == 1 && this->_joystick->GetRawButton(2) == 1) || (this->_joystick->GetRawButton(2) == 1 && this->_joystick->GetRawButton(1) == 1))
		return 0;
	else if (this->_joystick->GetRawButton(3) == 1)
		return 3;
	else if (this->_joystick->GetRawButton(2) == 1)
		return 2;
	else if (this->_joystick->GetRawButton(1) == 1)
		return 1;
	else
		return 0;
}

/*!
 \brief Get user-requested minibot shelf action (in/out).
 \details Read input from two thumb buttons on the joystick (buttons 4 and 5) and return ther states in a single variable.
 \return An integer with the last two bits being the right and left button inputs, respectively.
 */
int RJFRC2011::Controller::getMinibotSwitches()
{
	int switches = 0;
	// First bit is right switch value
	switches |= (this->_joystick->GetRawButton(4) << 0);
	// Second bit is left switch value
	switches |= (this->_joystick->GetRawButton(5) << 1);
	return switches;
}

/*!
 \brief A function that "normalizes" inputs from the joysticks (because they don't give perfect -1.0 to 1.0 values).
 \details If the requested value is negative, return its percentage of the minimum possible value; if it's possible, do the same with the max.  If it's zero, of course, return zero.
 \param joyVal the input from the joystick
 \param min the minimun joystick value
 \param max the maximum joystick value
 \return the normalized value
 \author Adam Bryant
 */
float RJFRC2011::Controller::normalize(float joyVal, float min, float max)
{
	float retVal = 0.0;
	if (joyVal < 0.0)
		retVal = abs(joyVal) / min;
	else if (joyVal > 0.0)
		retVal = abs(joyVal) / max;
	if (retVal < -1.0)
		retVal = -1.0;
	else if (retVal > 1.0)
		retVal = 1.0;
	return retVal;
}

/*!
 \brief An exponential function used to make joysticks less sensitive near the center and more sensitive towards the edges.
 \details Basically, plug the value requested by the user and a predefined constant into an exponential equation and return the result.
 \param x the value to be exponentiated
 \param a a predefined exponential factor
 \return the "expo-ed" value
 \author Adam Bryant
 */
float RJFRC2011::Controller::expo(float x, float a)
{
	return (a * (x * x * x) + (1 - a) * x);
}

/*!
 \brief Absolute value of a float, since I'm not sure if we can import the <cmath> library onto the cRIO.  EDIT: turns out we can, but I'll leave this here to conserve memory.
 \param initial the initial value
 \return the absolute value of the passed value; if it's negative, make it positive
 \author Matthew Haney
 */
float RJFRC2011::Controller::abs(float initial)
{
	if (initial < 0)
		return (initial * -1.0);
	else if (initial > 0)
		return initial;
	else 
		return 0.0;
}
